### PR TITLE
Feature/ts migration

### PR DESCRIPTION
**Pull request by maxdiebold-erg**
## Related Issues:
* Continuing the process of converting the project to TypeScript

## Main Changes:
* Migrated the _LocationSearch_ context, along with several other contexts, and updated them to export a wrapper function that internalizes the call to `useContext`.
* Migrated the _WaterbodyInfo_ file.
* Migrated the _MapMouseEvents_ file.
* Migrated several other shared components.

## Steps To Test:
1. Test that the project compiles without errors.
2. The changes in the _LocationSearch_ file are likely the most widespread, and I changed some of the state variables to make them more consistent (see the `FetchState` type in that file). I took care to ensure the appropriate changes were made in the corresponding variables project-wide, but more testing will be helpful until those files get converted to TS as well.

### CODE REVIEW

File: app/client/src/components/shared/AddDataWidget.SearchPanel.tsx [hunk 1]
```diff
@@ -740,59 +790,60 @@ function ResultCard({ result }: ResultCardProps) {
   /**
    * Adds non-tots layers as reference portal layers.
    */
-  function addRefLayer() {
+  const addRefLayer = useCallback(async () => {
```
Comment (cschwinderg) [inline app/client/src/components/shared/AddDataWidget.SearchPanel.tsx:793]:
```suggestion
  const addRefLayer = useCallback(() => {
```

I don't really like the idea of using async inside of a useCallback like this. Was there a specific need for this? 

The next 3 items relate to reverting this back to the way it was while still making TypeScript happy.

File: app/client/src/components/shared/AddDataWidget.SearchPanel.tsx [hunk 2]
```diff
@@ -740,59 +790,60 @@ function ResultCard({ result }: ResultCardProps) {
...
-              }
-            }
-          } else if (loadStatus === 'failed') {
-            setStatus('error');
+    const layer: Layer | __esri.TileLayer | __esri.GroupLayer =
+      await Layer.fromPortalItem({
+        portalItem: new PortalItem({
+          id: result.id,
+        }),
+      });
```
Comment (cschwinderg) [inline app/client/src/components/shared/AddDataWidget.SearchPanel.tsx:803]:
```suggestion
    Layer.fromPortalItem({
      portalItem: new PortalItem({
        id: result.id,
      }),
    }).then((layer) => {
```

Continuation of above.

File: app/client/src/components/shared/AddDataWidget.SearchPanel.tsx [hunk 3]
```diff
@@ -740,59 +790,60 @@ function ResultCard({ result }: ResultCardProps) {
...
-    });
-  }
+          // make all child layers visible, if applicable
+          if (isGroupLayer(layer)) {
+            layer.layers.forEach((tempLayer: Layer) => {
+              tempLayer.visible = true;
+            });
+          }
+          if ('sublayers' in layer) {
+            layer.sublayers.forEach((tempLayer: __esri.Sublayer) => {
```
Comment (cschwinderg) [inline app/client/src/components/shared/AddDataWidget.SearchPanel.tsx:829]:
```suggestion
            (layer as __esri.TileLayer).sublayers.forEach(
```
Continuation of above.

File: app/client/src/components/shared/AddDataWidget.SearchPanel.tsx [hunk 4]
```diff
@@ -740,59 +790,60 @@ function ResultCard({ result }: ResultCardProps) {
...
+    );
+
+    setWatcher(newWatcher);
+
+    // add the layer to the map
+    setWidgetLayers((currentWidgetLayers: WidgetLayer[]) => [
+      ...currentWidgetLayers,
+      layer,
+    ]);
+  }, [mapView, result.id, setWidgetLayers]);
```
Comment (cschwinderg) [inline app/client/src/components/shared/AddDataWidget.SearchPanel.tsx:846]:
```suggestion
      ]);
    });
  }, [mapView, result.id, setWidgetLayers]);
```
Continuation of above.

**Review by cschwinderg — CHANGES_REQUESTED** [review]
Thank you for working on this. Glad to see some progress on the TypeScript migration. I'm still reviewing but here are some initial issues that I found. I would still like to hear @courtneymyers input as well. 

Issues:
* I'm sometimes seeing infinite loading spinners when switching between locations on the community page.
* The plan summary page isn't working: http://localhost:3000/plan-summary/DOEE/40594
image
* Widgets are missing from the community landing page (i.e., http://localhost:3000/community).
image
* Seeing some console errors on the waterbody report page (i.e., http://localhost:3000/waterbody-report/21AWIC/AL03150110-0202-300/2020) and tribal page (i.e., http://localhost:3000/tribe/DELAWARENATION):
image
* Smaller items below.

**Comment by maxdiebold-erg** [comment]
@cschwinderg I forgot the `mapWidgets` file was a part of this PR, I just addressed that problem and I'll push the changes up: I'd factored out the `ShowUpstreamWatershed` and `ShowAllWaterbodies` components, then realized that widgets can't access context providers, so any context values need to be passed in (or be in their closure).

File: app/client/src/components/shared/AddDataWidget.SearchPanel.tsx (hunk 1 above)
Comment (maxdiebold-erg) [reply app/client/src/components/shared/AddDataWidget.SearchPanel.tsx:793]: No particular need, I'll revert that back.

**Comment by maxdiebold-erg** [comment]
I think migrating the `LocationSearch` context has too many far-reaching changes, and it should be a PR of its own. I'll open a new PR with the other changes, and save the `LocationSearch` changes for later.

**Pull request closed without merging**